core/src/test/scala/org/apache/spark/sql/delta/DeltaPurgeSuite.scala

```diff
@@ -0,0 +1,101 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta
+
+import org.apache.spark.sql.delta.test.DeltaSQLCommandTest
+import org.apache.spark.sql.QueryTest
+import org.apache.spark.sql.test.SharedSparkSession
+
+class DeltaPurgeSuite extends QueryTest
+  with SharedSparkSession
+  with DeltaSQLCommandTest
+  with DeletionVectorsTestUtils {
+
+  import testImplicits._
+
+  def executePurge(table: String, condition: Option[String] = None): Unit = {
+    condition match {
+      case Some(cond) => sql(s"REORG TABLE delta.`$table` WHERE $cond APPLY (PURGE)")
+      case None => sql(s"REORG TABLE delta.`$table` APPLY (PURGE)")
+    }
+  }
+
+  testWithDVs("Purge DVs will combine small files") {
+    withTempDir { tempDir =>
+      val path = tempDir.getCanonicalPath
+      val log = DeltaLog.forTable(spark, path)
+      spark
+        .range(0, 100, 1, numPartitions = 5)
+        .write
+        .format("delta")
+        .save(path)
+      sql(s"DELETE FROM delta.`$path` WHERE id IN (0, 99)")
+      assert(log.update().allFiles.filter(_.deletionVector != null).count() === 2)
+      executePurge(path)
+      val (addFiles, _) = getFileActionsInLastVersion(log)
+      assert(addFiles.forall(_.deletionVector === null))
+      checkAnswer(
+        sql(s"SELECT * FROM delta.`$path`"),
+        (1 to 98).toDF())
+    }
+  }
+
+  testWithDVs("Purge DVs") {
+    withTempDir { tempDir =>
+      val path = tempDir.getCanonicalPath
+      val log = DeltaLog.forTable(spark, path)
+      spark
+        .range(0, 100, 1, numPartitions = 5)
+        .write
+        .format("delta")
+        .save(path)
+      sql(s"DELETE FROM delta.`$path` WHERE id IN (0, 99)")
+      assert(log.update().allFiles.filter(_.deletionVector != null).count() === 2)
+
+      // First purge
+      executePurge(path)
+      val (addFiles, _) = getFileActionsInLastVersion(log)
+      assert(addFiles.size === 1) // two files are combined
+      assert(addFiles.forall(_.deletionVector === null))
+      checkAnswer(
+        sql(s"SELECT * FROM delta.`$path`"),
+        (1 to 98).toDF())
+
+      // Second purge is a noop
+      val versionBefore = log.update().version
+      executePurge(path)
+      val versionAfter = log.update().version
+      assert(versionBefore === versionAfter)
+    }
+  }
+
+  test("Purge a non-DV table is a noop") {
+    withTempDir { tempDir =>
+      val path = tempDir.getCanonicalPath
+      val log = DeltaLog.forTable(spark, path)
+      spark
+        .range(0, 100, 1, numPartitions = 5)
+        .write
+        .format("delta")
+        .save(path)
+      val versionBefore = log.update().version
+      executePurge(path)
+      val versionAfter = log.update().version
+      assert(versionBefore === versionAfter)
+    }
+  }
```

Comment:
We may need one or two tests for the case where reorg is run on a set of partitions.

Reply:
Done!